MaterialDesignThemes.Wpf/DataGridAssist.cs

```diff
@@ -1,281 +1,258 @@
+            var dataGrid = (DataGrid)sender;
+
+            if (e.Key == Key.Space && e.OriginalSource is DataGridCell cell
+                && !cell.IsReadOnly && cell.Column is DataGridComboBoxColumn)
             {
                 dataGrid.BeginEdit();
             }
         }
 
-        private static void DataGridOnPreviewMouseLeftButtonDown(object sender,
-            MouseButtonEventArgs mouseButtonEventArgs)
+        /// <summary>
+        /// Allows editing of components inside of a datagrid cell with a single left click.
+        /// </summary>
+        private static void AllowDirectEditWithoutFocus(object sender, MouseButtonEventArgs mouseArgs)
         {
-            var dataGrid = (DataGrid)sender;
-
-            var inputHitTest =
-                dataGrid.InputHitTest(mouseButtonEventArgs.GetPosition((DataGrid)sender)) as DependencyObject;
+            var originalSource = (DependencyObject)mouseArgs.OriginalSource;
+            var dataGridCell = originalSource
+                .GetVisualAncestry()
+                .OfType<DataGridCell>()
+                .FirstOrDefault();
+
+            // Readonly has to be handled as the passthrough ignores the
+            // cell and interacts directly with the content
+            if (dataGridCell?.IsReadOnly ?? true)
+            {
+                return;
+            }
 
-            while (inputHitTest != null)
+            if (dataGridCell?.Content is UIElement element)
             {
-                var dataGridCell = inputHitTest as DataGridCell;
-                if (dataGridCell != null && dataGrid.Equals(dataGridCell.GetVisualAncestry().OfType<DataGrid>().FirstOrDefault()))
+                var dataGrid = (DataGrid)sender;
+                
+                // Check if the cursor actually hit the element and not just the cell
+                var mousePosition = mouseArgs.GetPosition(element);
+                var elementHitBox = new Rect(element.RenderSize);
+                if (elementHitBox.Contains(mousePosition))
                 {
-                    if (dataGridCell.IsReadOnly) return;
+                    dataGrid.CurrentCell = new DataGridCellInfo(dataGridCell);
+                    dataGrid.BeginEdit();
 
-                    ToggleButton toggleButton;
-                    ComboBox comboBox;
-                    if (IsDirectHitOnEditComponent(dataGridCell, mouseButtonEventArgs, out toggleButton))
+                    switch (dataGridCell?.Content)
                     {
-                        dataGrid.CurrentCell = new DataGridCellInfo(dataGridCell);
-                        dataGrid.BeginEdit();
-                        toggleButton.SetCurrentValue(ToggleButton.IsCheckedProperty, !toggleButton.IsChecked);
-                        dataGrid.CommitEdit();
-                        mouseButtonEventArgs.Handled = true;
+                        // Send a 'left click' routed command to the checkbox
+                        case CheckBox checkBox:
```

Comment:
Should this be `ToggleButton` rather than `CheckBox` (its base class). This is what the previous code was doing.

Reply:
Agreed.